src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java

```diff
@@ -1358,6 +1360,63 @@ private void preparePatternMatchingCatchIfNeeded(JCBlock tree) {
         }
     }
 
+    @Override
+    public void visitDerivedInstance(JCDerivedInstance tree) {
+        ListBuffer<JCStatement> newBlock = new ListBuffer<>();
+        VarSymbol temp;
+        if (tree.expr instanceof JCIdent i &&
+            i.sym.kind == Kind.VAR &&
+            (i.sym.owner.kind == Kind.MTH || i.sym.owner.kind == Kind.VAR)) {
+            temp = (VarSymbol) i.sym;
+        } else {
+            temp = new VarSymbol(Flags.SYNTHETIC,
+                    names.fromString("expr" + variableIndex++ + target.syntheticNameChar() + "temp"),
+                    tree.expr.type,
+                    currentMethodSym);
+            newBlock.add(make.VarDef(temp, translate(tree.expr)));
+        }
+
+
+        ClassSymbol recordClass = (ClassSymbol) tree.expr.type.tsym;
+        List<VarSymbol> outgoingBindingsIt = tree.componentLocalVariables;
+        List<? extends RecordComponent> recordComponentsIt = recordClass.getRecordComponents();
+
+        while (outgoingBindingsIt.nonEmpty()) {
+            Type erasedComponentType = types.erasure(recordComponentsIt.head.type);
+            newBlock.add(make.VarDef(outgoingBindingsIt.head,
+                                     make.App(make.Select(make.Ident(temp),
+                                                          recordComponentsIt.head.accessor))
+                                         .setType(erasedComponentType)));
+            outgoingBindingsIt = outgoingBindingsIt.tail;
+            recordComponentsIt = recordComponentsIt.tail;
+        }
+
+        newBlock.add(translate(tree.block));
+
+        JCNewClass createNew = make.NewClass(null,
+                                             List.nil(),
+                                             make.QualIdent(recordClass),
+                                             tree.componentLocalVariables.map(make::Ident),
+                                             null);
+
+        createNew.type = tree.type;
+
+        List<Type> canonicalConstructorTypes =
+                recordClass.getRecordComponents()
+                           .stream()
+                           .map(c -> types.erasure(c.type))
+                           .collect(List.collector());
+        MethodSymbol init = rs.resolveInternalMethod(tree.pos(),
+                                                     env,
+                                                     tree.type,
+                                                     names.init,
+                                                     canonicalConstructorTypes,
+                                                     List.nil());
+        createNew.constructor = init;
```

Comment:
Maybe instead of hardcoding the constructor that is canonical at compile time (in case new components get added), it might be better to go through some sort of `indy` callsite like (at least when not in the same compilation unit):
```
DynamicCallSiteDesc[java.lang.runtime.RecordSupport::derivedConstructor(modified component names...):(T extends Record, Class<?>... /* modified component types */)T]
```

with the `derivedConstructor` bootstrap method:
```java
public static MethodHandle derivedConstructor(MethodHandles.Lookup lookup, String unused, MethodType type, String... modifiedComponents) throws ReflectiveOperationException {
	requireNonNull(lookup);
	requireNonNull(type);
	// implicit null-check:
	List<String> modifiedComponentNames = List.of(modifiedComponents);

	Class<?> rtype = type.returnType();
	if (
		!rtype.isRecord()
		|| type.parameterCount() != modifiedComponents.length + 1
		|| type.parameterType(0) != rtype
	) {
		throw new IllegalArgumentException("...");
	}

	Set<String> remainingComponentNames = new HashSet(modifiedComponentNames);
	if (remainingComponentNames.size() != modifiedComponentNames.size()) {
		throw new IllegalArgumentException("Duplicate component names in modifiedComponents");
	}

	RecordComponent[] recordComponents = rtype.getRecordComponents();

	var componentTypes = new Class<?>[recordComponents.length];
	var filters = new MethodHandle[recordComponents.length];
	var reorder = new int[recordComponents.length];

	for (int i = 0, j = 1; i < recordComponents.length; i++) {
		var component = recordComponents[i];
		componentTypes[i] = component.getType();

		var getter = lookup.findVirtual(rtype, component.getName(), MethodType.methodType(component.getType()));
		if (modifiedComponentNames.contains(component.getName())) {
			remainingComponentNames.remove(component.getName());
			filters[i] = null;
			reorder[i] = j++;
		} else {
			filters[i] = getter;
			reorder[i] = 0;
		}
	}

	if (!remainingComponentNames.isEmpty()) {
		throw new IllegalArgumentException("Components " + remainingComponentNames + " are not present in the record " + rtype);
	}

	var canonicalConstructor = lookup.findConstructor(rtype, MethodType.methodType(rtype, componentTypes);
	var filteredConstructor = MethodHandles.filterArguments(canonicalConstructor, 0, filters);
	// implicitly verifies that type's parameter types match the actual component types
	var permutedConstructor = MethodHandles.permuteArguments(filterArguments, type, reorder);

	return permutedConstructor;
}
```

Reply:
While this is very smart, I would prefer not to do that. JLS 13.4.27 talks about binary compatibility for record, and warns that adding or removing components may break record usages. So I don't think we need to provide too much support for adding or removing record components.

Overall, I believe the current code adheres to what the JLS draft says. If the author of the record wants to add or remove components, they can make it so that the existing compiled code still works (by adding constructor overloads and keeping the accessors for removed component). It is better, IMO, to keep the behavior predictable, and leave adjustments after record component modifications up to the record author, than to introduce some magic.

Reply:
This was once again asked about on amber‑dev:
- https://mail.openjdk.org/pipermail/amber-dev/2024-April/008748.html